### PR TITLE
fix(product): Support non-Latin characters in product handles with Un…

### LINKED ISSUE #14378: [Bug]: Admin cannot create products with Arabic titles (invalid handle)
### Package.json file

```JSON
{}
```

### Node.js version

v20.19.5

### Database and its version

PostgreSQL 17.6

### Operating system name and version

Windows 11

### Browser name

_No response_

### What happended?

When creating products, collections, or categories from Medusa Admin using Arabic titles, the backend throws an error and creation fails.

**Error**
`Invalid product handle '-'. It must contain URL safe characters`

**Workaround**
Using `slugify` to generate a handle based on the entered title works.

https://github.com/user-attachments/assets/529dacc4-7fe8-4396-8640-b92fbc508371

### Expected behavior

Entities should be created successfully with non-Latin (Arabic) titles.

### Actual behavior

Handle generation strips Arabic characters, resulting in an invalid handle (-) and the request fails.

### Link to reproduction repo

https://github.com/medusajs/medusa

### PR DESCRIPTION
**Pull request by Cpidar**
# Summary (final PR implementation)

**What** — What changes are introduced in this PR?

This PR fixes and closes issue #14378: products with non-Latin titles (Persian, Arabic, Japanese, Chinese, Korean, Russian, etc.) cannot be created because handle generation fails.

**Three specific changes:**

1. **Improve `toHandle()`** to preserve Unicode letters from any language while still removing special characters (apostrophes, punctuation, etc.)
2. **Update `isValidHandle()`** to be Unicode-aware with proper lowercase validation
3. **Add comprehensive tests** for all scripts and edge cases

**Why** — Why are these changes relevant or necessary?  

Currently, products with non-Latin titles (Persian, Arabic, Japanese, Chinese, Korean, Russian, etc.) cannot be created. The error message is:
```
Invalid product handle '-'. It must contain URL safe characters
```

Root cause: `toHandle()` was stripping all non-ASCII characters, resulting in empty or partial handles.

**How** — How have these changes been implemented?

**Change 1: Improve `toHandle()` in `to-handle.ts`**
- Preserves Unicode letters (`\p{L}`) and numbers (`\p{N}`)
- Removes special characters (apostrophes, punctuation, symbols)
- Replaces spaces/underscores with hyphens
- Falls back to random suffix for empty results

**Change 2: Update `isValidHandle()` in `is-valid-handle.ts`**
- Rejects uppercase ASCII letters (maintains backward compatibility)
- Accepts lowercase letters + case-less scripts (Arabic, Persian, CJK)
- Validates format (no spaces, no special chars, valid hyphen placement)

***Note: No change to product module*** - continues using `toHandle()`

**Testing** — How have these changes been tested, or how can the reviewer test the feature?

- Added unit tests for 10+ language scripts (Persian, Arabic, Japanese, Chinese, Korean, Russian, English)
- Added edge case tests (empty strings, special characters only, accents)
- Verified backward compatibility (Latin handles unchanged)
- Verified categories/collections unchanged

---

## Examples

```typescript
// Persian
toHandle("کتاب فارسی") // "کتاب-فارسی"

// Accents preserved
toHandle("café") // "café"

// Special characters removed
toHandle("Women's Shoes") // "womens-shoes"

// Edge case fallback
toHandle("") // "product-abc123" (random, unique)

// Validation
isValidHandle("کتاب-فارسی") // true
isValidHandle("Hello-World") // false (uppercase rejected)
```

---

## Checklist

Please ensure the following before requesting a review:

- [x] I have added a **changeset** for this PR
- [x] The changes are covered by relevant **tests**
- [x] I have verified the code works as intended locally
- [x] Verified backwards compatibility
- [x] I have linked the related issue(s) if applicable

---

## Additional Context

Closes #14378

This approach follows the established pattern of major ecommerce platforms:

| Platform | Slug/Handle Approach | Unicode Support |
|----------|---------------------|-----------------|
| **WordPress** | `sanitize_title()` – preserves Unicode | ✅ Yes |
| **WooCommerce** | UTF-8 in product permalinks | ✅ Yes |
| **Shopify** | Transliterates to Latin or preserves Unicode | ✅ Yes |
| **Shopware** | Full Unicode slug support | ✅ Yes |
| **Medusa (Current)** | Latin-only + stripping | ❌ No |
| **Medusa (This PR)** | Unicode-aware + `kebabCase` | ✅ Yes |

### CODE REVIEW

**Comment by Cpidar** [comment]
@medusa-os-bot Thank you for the feedback! I've addressed all issues:

✅ **Changeset format fixed** - Updated to `fix(product, utils): Support non-Latin characters in product handles`

✅ **Tests added** - Created comprehensive tests for both `toHandle` and `isValidHandle` covering:
- Persian, Arabic, Japanese, Chinese, Korean, Russian scripts
- Special character removal (apostrophes, punctuation)
- Uppercase/lowercase handling (backward compatible)
- Empty handle fallback

✅ **Uppercase Latin handles fixed** - Added `.toLowerCase()` before validation to maintain backward compatibility

✅ **Special characters in kebabCase** - Kept `toHandle` function (not replaced with kebabCase) but improved it to handle non-latic characters while still removing special chars like apostrophes

✅ **Trailing newlines added** - Fixed formatting in changeset and validate-handle files

So now:
- ✅ `toHandle` is preserved (backward compatible)
- ✅ `toHandle` now properly handles Persian/Arabic/Japanese/etc. characters
- ✅ `toHandle` still removes special characters (apostrophes, `!@#$%`, etc.)
- ✅ `isValidHandle` is Unicode-aware with backward compatible lowercase normalization

The changes are now ready for review. All tests pass locally. Thank you for your patience with my first contribution!

**Comment by Cpidar** [comment]
**Update:** I've reverted the product module back to using `toHandle` instead of `kebabCase`.

**Comment by Cpidar** [comment]
@medusa-os-bot Thank you for the detailed review! I've addressed all remaining issues:

✅ **Quotes fixed** - Changed all single quotes to double quotes in `to-handle.ts` to match Prettier/Medusa standards

✅ **JSDoc updated** - Rewrote the comment to accurately reflect the new implementation (preserves Unicode, removes special chars, fallback behavior)

✅ **Uppercase handling fixed** - Removed `.toLowerCase()` from `isValidHandle` and now explicitly reject uppercase ASCII letters (A-Z). This ensures:
- Backward compatibility (uppercase handles were invalid before)
- No case-sensitive duplicate issues with the unique database index
- Auto-generated handles from `toHandle` are always lowercase, so this maintains consistency

✅ **Test added** - Added test case for uppercase rejection

✅ **Test fix pushed:** Updated `to-handle.spec.ts` to expect fallback handles for edge cases (`""`, `"-"`, `"   "`, `"___"`, `"@#$%"`).

The fallback behavior is intentional - it guarantees every product gets a valid, unique handle rather than failing validation. This matches how WordPress, Shopify, and other platforms handle edge cases.

All tests now pass locally. ✅

The PR should now be ready for final review. Thank you for your patience and thorough feedback on my first contribution!

**Comment by Cpidar** [comment]
@medusa-os-bot Thank you for the precise review! I've fixed the remaining issues:

✅ **JSDoc corrected** - Updated to accurately describe that `toHandle` preserves accented characters (é, ñ, ü, etc.) rather than stripping them.

✅ **Trailing newline added** - Added to `is-valid-handle.spec.ts`

✅ **Duplicate test block removed** - Cleaned up the duplicate "English / Latin - valid cases" section

The documentation now accurately reflects the implementation. Ready for final review!

**Comment by Cpidar** [comment]
I've fixed the remaining issue: Restore the trailing comma after toHandle in the import list.

**Comment by Cpidar** [comment]
@medusa-os-bot Formatting fixes applied:

✅ Added trailing comma to last array item in test expectations
✅ Removed extra blank lines in toHandle function body

The PR should now meet all Medusa code style conventions. 

Thank you for the thorough review - these formatting details are noted for future contributions.

Regarding the non-deterministic fallback: Yes, this is intentional to ensure unique handles for edge cases. Any tests for that path should use regex matching (e.g., `/^product-[a-z0-9]{6}$/`) rather than exact string comparison, as documented in the test file.

Ready for final review! 🚀

File: packages/core/utils/src/common/__tests__/to-handle.spec.ts [hunk 1]
```diff
@@ -1,61 +1,260 @@
...
+      "你好-世界",
+      "안녕하세요-세계",
+      "привет-мир",
+      "αβγ-δεζ",
+      "a",
+      "1",
+      "a-1-b-2",
+    ]
+
+    handles.forEach(handle => {
```
Comment (shahednasser) [inline packages/core/utils/src/common/__tests__/to-handle.spec.ts]: lint error here

File: packages/core/utils/src/common/to-handle.ts [hunk 2]
```diff
@@ -1,20 +1,24 @@
...
-    .replace(/[^a-z0-9A-Z-]/g, "")
-    .replace(/-{2,}/g, "-")
+  let handle = value
+    .toLowerCase()                          // Normalize to lowercase
+    .replace(/ß/g, "ss")                    // Convert German eszett to ss
+    .replace(/[^\p{L}\p{N}\s_-]/gu, "")     // Remove all characters except Unicode letters, numbers, spaces, underscores, and hyphens
+    .replace(/[\s_]+/g, "-")                // Replace spaces and underscores with single hyphens
+    .replace(/-+/g, "-")                    // Collapse multiple consecutive hyphens into one
+    .replace(/^-|-$/g, "")                  // Trim leading and trailing hyphens
+  if (!handle) {
```
Comment (shahednasser) [inline packages/core/utils/src/common/to-handle.ts:21]: fix lint errors

File: packages/core/utils/src/common/validate-handle.ts [hunk 3]
```diff
@@ -1,7 +1,41 @@
...
+ * isValidHandle("کتاب-فارسی") // true
+ * isValidHandle("私の-製品") // true
+ * isValidHandle("我的-产品") // true
+ * isValidHandle("Hello-World") // false (contains uppercase)
+ * isValidHandle("my product") // false (spaces not allowed)
+ * isValidHandle("-product") // false (leading hyphen)
  */
 export const isValidHandle = (value: string): boolean => {
-  return /^[a-z0-9]+(?:-[a-z0-9]+)*$/.test(value)
+  if (!value || value.length === 0) return false
```
Comment (shahednasser) [inline packages/core/utils/src/common/validate-handle.ts]: fix lint error

Comment (shahednasser) [reply packages/core/utils/src/common/validate-handle.ts]: Also can be simplfied to `!value?.length`

**Review by shahednasser — CHANGES_REQUESTED** [review]
Thank you for your contribution! Works as expected, but had some minor comments

**Comment by Cpidar** [comment]
Could you please clarify the lint errors? I'm not seeing the exact issue from the comment alone or in my IDE. What rule is being violated? 
Thank you for your patience with my first contribution! 🙏

**Comment by Cpidar** [comment]
> Thank you for your contribution! Works as expected, but had some minor comments

I can't see the exact lint error from CI. Based on common lint patterns, I've made these changes:

- Removed inline comments in `to-handle.ts`
- Added empty line before line 13 in `to-handle.ts`
- Changed `validate-handle.ts` line 27 to use optional chaining: `if (!value?.length) return false` + added empty line
- Changed `to-handle.spec.ts` line 209 to `handles.forEach((handle) => {`

If this doesn't resolve it, please share the specific error message. Thanks!

**Comment by Cpidar** [comment]
Prettier formatting applied to all files (`yarn prettier --write`).

- Removed blank lines after `{`
- Removed trailing whitespace
- Standardized indentation

CI should pass now.

**Comment by Cpidar** [comment]
**Fix applied:**

✅ Added `\p{Lm}` (modifier letters) to `isValidHandle` regex

This allows Japanese prolonged sound mark `ー` (U+30FC) and iteration mark `々` (U+3005) to pass validation.

**Changes made:**

| File | Change |
|------|--------|
| `validate-handle.ts` | Added `\p{Lm}` to character class |
| `to-handle.spec.ts` | Added invariant test + Japanese test cases |

**New test cases added:**
- `ラーメン` (ramen)
- `様々` (various)
- `コーヒー` (coffee)
- `サーバー` (server)
- `人々` (people)
- `時々` (sometimes)

**Invariant test added:**
```typescript
expect(isValidHandle(toHandle(title))).toBe(true)

**Comment by Cpidar** [comment]
Thank you for the review and guidance! 🙏
Looking forward to contributing more to Medusa in the future!
